### ISSUE TITLE
Protocols as function arguments in protocols are not recognized

### ISSUE THREAD

**Issue by jmidt** [python/mypy#12644]
**Bug Report**

A Protocol (A) that involves a function with another Protocol (B) as an argument is not seen as compatible with a function that takes an argument that correctly fulfills Protocol B. 

**To Reproduce**
This will fail on mypy 0.940:
```python
from typing import Protocol

class AbstractArgument(Protocol):
    pass

class ConcreteArgument:
    pass

class TakesAbstract(Protocol):
    def __call__(self, arg: AbstractArgument) -> None:
        pass

def concrete_func(arg: ConcreteArgument) -> None:
    pass

# Mypy does not complain - this cast works fine
foo: AbstractArgument = ConcreteArgument()

# Mypy complains:
# "Incompatible types in assignment (expression has type "Callable[[ConcreteArgument], None]", 
# variable has type "TakesAbstract")"
test_func: TakesAbstract = concrete_func
```

**Expected Behavior**

test_func should be correctly cast to a `TakesAbstract`. In this simple case, a `Callable[...]` would suffice, but if the protocol `TakesAbstract` grows more complicated, this is not the case any more and the example still fails. 

**Actual Behavior**

mypy raises an error:
`Incompatible types in assignment (expression has type "Callable[[ConcreteArgument], None]", variable has type "TakesAbstract")`

**Your Environment**

Python 3.9.5
mypy 0.940 with --strict

**Comment by JelleZijlstra** [comment]
Closing per @erictraut's explanation. (Thanks Eric!)
